architecture/spmc.rst

```diff
@@ -0,0 +1,322 @@
+                autoactivate off
+                spmc_sp_handler -> SP: __thread_enter_user_mode \n <font color=red>ERET
+                activate SP
+
+                SP -> spmc_sp_handler : (FF-A message) sp_handle_svc \n <font color=red>SVC
+                deactivate SP
+                activate spmc_sp_handler
+                spmc_sp_handler -> spmc_sp_handler: Store SP context
+                spmc_sp_handler -> spmc_sp_handler: return to sp_enter_invoke_cmd
+                deactivate spmc_sp_handler
+                spmc_sp_handler -> spmc_sp_handler: Retrieve FF-A message from SP context
+                return
+                return
+                return
+        end
+                return
+        == End of kernel thread ==
+                return
+        return \n <font color=red>SMC
+        return FFA_MSG_SEND_DIRECT_RESP \n <font color=red>ERET
+
+
+Every message received by the SPMC from the Normal World is handled in the
+``thread_spmc_msg_recv`` function.
+
+When entering a SPs we need to be running on a kernel thread. This is needed to
+be able to push the TS session (We push the TS session to get access to the SP
+memory).
+Currently the only possibility to enter a SP from the Normal world is via a
+``FFA_MSG_SEND_DIRECT_REQ``. Whenever we receive a FFA_MSG_SEND_DIRECT_REQ
+message which doesn't have OP-TEE as the endpoint-id, we start a kernel thread
+and forward the FF-A message to  the ``sp_msg_handler``.
+
+The sp_msg_handler is responsible for all messages coming or going to/from a SP.
+It runs in a while loop and will handle every message until it comes across a
+messages which is not intended for the secure world.
+After a message is handled by the SPMC or when it needs to be forwarded to a SP,
+sp_enter is called.
+sp_enter() will copy the FF-A arguments and resume the SP.
+
+When the SPMC needs to have access to the SPs memory, it will call
+ts_push_current_session to gain access and ts_pop_current_session to release the
+access.
+
+Running and exiting SPs
+-----------------------
+
+THE SPMC resumes/starts q SP by calling the sp_enter(). This will set-up the SP
+context and jump into S-EL0.
+Whenever the SP performs a system call it will end-up in sp_handle_svc().
+sp_handle_svc() store the current context of the SP and makes sure that we don't
+return to S-EL0 but instead returns to S-EL1 back to sp_enter(). sp_enter() will
+pass the FF-A registers (x0-x7) to spmc_sp_msg_handler(). This will process the
+FF-A message.
+
+
+RxTx buffer managment
+---------------------
+RxTx buffers are used by the SPMC to exchange information between an endpoin
+t and the SPMC. The rxtx_buf struct is used by the SPMC for abstracting buffer
+management.
+Every SP has a rxtx_buf struct wich will be passed to every function that needs
+access to the rxtx buffer.
+A seperate struct rxtx_buf is defined for the Normal World, which gives acces
+s to the Normal World buffers.
+
+
+Structs and global objects
```

Comment:
I don't think this section is useful at all. It is implementation details that will likely become outdated and which would be much easier documented and understood in the code itself.

Reply:
Done